### DISCUSSION TITLE
How to best cache a task based on input file content?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [PrefectHQ/prefect#15371] [Q&A]
Assume I have a task that takes as inputs an integer and a path to a file:

```python
from pathlib import Path

from prefect import task

@task
def read_line(line_number: int, file_path: Path) -> str:
    with open(file_path, "r") as f:
        lines = f.readlines()
    return lines[line_number]
```

The way I understand prefect's [cache key policies](https://docs.prefect.io/3.0/develop/task-caching#cache-policies), by default, a cache key for the `read_line` task will be computed based on the values of the input arguments. Thus, if I call the task a second time with the same arguments, but the content of the file at `file_path` has changed in the meantime, the task will mistakenly not run again.

Does prefect support generating cache keys based on the contents of files and directories referred by `pathlib.Path` objects? If not, I guess I can write a [cache key function](https://docs.prefect.io/3.0/develop/task-caching#cache-key-functions) that replaces the values of `pathlib.Path` arguments with a hash of their contents, and then calls the original cache key function. Where can I find the latter? Are there some pitfalls I am unaware of?

**Comment by desertaxle** [comment]
Hey @NiklasKappel! Writing a cache key function is the suggested way to cache based on the contents of a file or directory. Be sure to handle cases where a file or directory doesn't exist in your cache key function. Exceptions in a cache key function will prevent your tasks from executing.

**Comment by jakob1379** [comment]
adding this natively in Prefect would make choosing Prefect over snakemake/nextflow a no-brainer.
